File: examples/clients/javascript/websocket-client.js
/**
 * Songbird WebSocket Client - Real-Time Communication
 * 
 * Official JavaScript/Node.js client for Songbird WebSocket API.
 * Provides real-time bidirectional communication with event subscriptions.
 * 
 * Version: 0.2.1
 * Last Updated: November 11, 2025 - Phase 4
 * 
 * Features:
 * - Real-time bidirectional communication
 * - Event subscription system
 * - Automatic reconnection
 * - Ping/pong keep-alive
 * - Query federation status
 * - Service discovery
 * - Full error handling
 * 
 * Requirements:
 *     npm install ws
 * 
 * Usage:
 *     // Basic connection
 *     const client = new SongbirdWebSocketClient('ws://localhost:8080/api/ws/ws');
 *     await client.connect();
 *     
 *     // Subscribe to events
 *     await client.subscribe(['service_update', 'health_update']);
 *     
 *     // Query status
 *     const status = await client.queryStatus();
 *     
 *     // Listen for events
 *     client.on('service_update', (event) => {
 *         console.log('Service updated:', event);
 *     });
 */

const WebSocket = require('ws');
const EventEmitter = require('events');

/**
 * Federation Status
 */
class FederationStatus {
    constructor(data) {
        this.totalServices = data.total_services;
        this.totalPeers = data.total_peers;
        this.uptimeSeconds = data.uptime_seconds;
    }
}

/**
 * Service Summary
 */
class ServiceSummary {
    constructor(data) {
        this.name = data.name;
        this.address = data.address;
        this.port = data.port;
        this.capabilities = data.capabilities;
    }
}

/**
 * Songbird WebSocket Client
 * 
 * Provides real-time bidirectional communication with Songbird.
 * 
 * @extends EventEmitter
 * 
 * @example
 * const client = new SongbirdWebSocketClient('ws://localhost:8080/api/ws/ws');
 * await client.connect();
 * 
 * // Subscribe to events
 * await client.subscribe(['service_update', 'health_update']);
 * 
 * // Listen for events
 * client.on('service_update', (event) => {
 *     console.log('Service updated:', event.service_name);
 * });
 * 
 * // Query status
 * const status = await client.queryStatus();
 * console.log(`Services: ${status.totalServices}`);
 * 
 * await client.close();
 */
class SongbirdWebSocketClient extends EventEmitter {
    /**
     * Create a WebSocket client
     * 
     * @param {string} url - WebSocket URL (e.g., 'ws://localhost:8080/api/ws/ws')
     * @param {Object} options - Client options
     * @param {number} options.pingInterval - Seconds between ping messages (default: 30)
     * @param {number} options.pingTimeout - Timeout for pong response (default: 10)
     * @param {boolean} options.autoReconnect - Automatically reconnect (default: true)
     * @param {number} options.maxReconnectAttempts - Max reconnection attempts (default: 5)
     */
    constructor(url, options = {}) {
        super();
        
        this.url = url;
        this.pingInterval = (options.pingInterval || 30) * 1000;
        this.pingTimeout = (options.pingTimeout || 10) * 1000;
        this.autoReconnect = options.autoReconnect !== false;
        this.maxReconnectAttempts = options.maxReconnectAttempts || 5;
        
        this.ws = null;
        this.connected = false;
        this.pingTimer = null;
        this.subscriptions = [];
        this.reconnectAttempts = 0;
        this.pendingRequests = new Map();
        this.requestId = 0;
    }
    
    /**
     * Connect to Songbird WebSocket server
     * 
     * @returns {Promise<void>}
     * @throws {Error} If connection fails
     */
    connect() {
        return new Promise((resolve, reject) => {
            console.log(`Connecting to ${this.url}`);
            
            this.ws = new WebSocket(this.url);
            
            this.ws.on('open', () => {
                this.connected = true;
                this.reconnectAttempts = 0;
                console.log('Connected successfully');
                
                // Start ping loop
                this._startPingLoop();
                
                this.emit('connected');
                resolve();
            });
            
            this.ws.on('message', (data) => {
                try {
                    const message = JSON.parse(data.toString());
                    this._handleMessage(message);
                } catch (error) {
                    console.error('Failed to parse message:', error);
                }
            });
            
            this.ws.on('close', () => {
                this.connected = false;
                this._stopPingLoop();
                console.log('Connection closed');
                
                this.emit('disconnected');
                
                if (this.autoReconnect) {
                    this._reconnect();
                }
            });
            
            this.ws.on('error', (error) => {
                console.error('WebSocket error:', error);
                this.emit('error', error);
                
                if (!this.connected) {
                    reject(error);
                }
            });
        });
    }
    
    /**
     * Close WebSocket connection
     * 
     * @returns {Promise<void>}
     */
    async close() {
        this.autoReconnect = false;
        this._stopPingLoop();
        
        if (this.ws) {
            this.ws.close();
        }
        
        console.log('Connection closed');
    }
    
    /**
     * Start ping loop
     * @private
     */
    _startPingLoop() {
        this.pingTimer = setInterval(() => {
            this.ping().catch(err => {
                console.warn('Ping failed:', err);
            });
        }, this.pingInterval);
    }
    
    /**
     * Stop ping loop
     * @private
     */
    _stopPingLoop() {
        if (this.pingTimer) {
            clearInterval(this.pingTimer);
            this.pingTimer = null;
        }
    }
    
    /**
     * Handle reconnection
     * @private
     */
    async _reconnect() {
        if (this.reconnectAttempts >= this.maxReconnectAttempts) {
            console.error('Max reconnection attempts reached');
            return;
        }
        
        this.reconnectAttempts++;
        const delay = Math.min(Math.pow(2, this.reconnectAttempts) * 1000, 30000);
        
        console.log(`Reconnecting in ${delay/1000}s (attempt ${this.reconnectAttempts})`);
        
        await new Promise(resolve => setTimeout(resolve, delay));
        
        try {
            await this.connect();
            
            // Re-subscribe to events
            if (this.subscriptions.length > 0) {
                await this.subscribe(this.subscriptions);
            }
        } catch (error) {
            console.error('Reconnection failed:', error);
            this._reconnect();
        }
    }
    
    /**
     * Send message to server
     * @private
     */
    _sendMessage(message) {
        if (!this.connected || !this.ws) {
            throw new Error('Not connected');
        }
        
        this.ws.send(JSON.stringify(message));
    }
    
    /**
     * Handle incoming message
     * @private
     */
    _handleMessage(message) {
        const { type } = message;
        
        // Emit event for message type
        this.emit(type, message);
        
        // Handle responses to requests
        if (message.requestId !== undefined) {
            const pending = this.pendingRequests.get(message.requestId);
            if (pending) {
                pending.resolve(message);
                this.pendingRequests.delete(message.requestId);
            }
        }
    }
    
    /**
     * Send request and wait for response
     * @private
     */
    _sendRequest(message, timeout = 5000) {
        return new Promise((resolve, reject) => {
            const requestId = this.requestId++;
            message.requestId = requestId;
            
            const timer = setTimeout(() => {
                this.pendingRequests.delete(requestId);
                reject(new Error('Request timeout'));
            }, timeout);
            
            this.pendingRequests.set(requestId, {
                resolve: (response) => {
                    clearTimeout(timer);
                    resolve(response);
                },
                reject
            });
            
            this._sendMessage(message);
        });
    }
    
    /**
     * Subscribe to event types
     * 
     * @param {string[]} events - Event types to subscribe to
     * @returns {Promise<void>}
     * 
     * @example
     * await client.subscribe(['service_update', 'health_update']);
     */
    async subscribe(events) {
        this.subscriptions.push(...events);
        
        this._sendMessage({
            type: 'subscribe',
            events: events
        });
        
        console.log('Subscribed to:', events);
    }
    
    /**
     * Unsubscribe from event types
     * 
     * @param {string[]} events - Event types to unsubscribe from
     * @returns {Promise<void>}
     */
    async unsubscribe(events) {
        this.subscriptions = this.subscriptions.filter(e => !events.includes(e));
        
        this._sendMessage({
            type: 'unsubscribe',
            events: events
        });
        
        console.log('Unsubscribed from:', events);
    }
    
    /**
     * Send ping message
     * 
     * @param {string} [data] - Optional data to include
     * @returns {Promise<void>}
     */
    async ping(data = null) {
        const message = { type: 'ping' };
        if (data) {
            message.data = data;
        }
        this._sendMessage(message);
    }
    
    /**
     * Query federation status
     * 
     * @returns {Promise<FederationStatus>}
     * 
     * @example
     * const status = await client.queryStatus();
     * console.log(`Services: ${status.totalServices}`);
     */
    async queryStatus() {
        this._sendMessage({ type: 'query_status' });
        
        return new Promise((resolve) => {
            const handler = (message) => {
                if (message.type === 'federation_status') {
                    this.removeListener('federation_status', handler);
                    resolve(new FederationStatus(message));
                }
            };
            this.on('federation_status', handler);
        });
    }
    
    /**
     * Query services by capability
     * 
     * @param {string[]} capabilities - Required capabilities
     * @returns {Promise<ServiceSummary[]>}
     * 
     * @example
     * const services = await client.queryServices(['ml', 'training']);
     * services.forEach(s => console.log(`Service: ${s.name}`));
     */
    async queryServices(capabilities) {
        this._sendMessage({
            type: 'query_services',
            capabilities: capabilities
        });
        
        return new Promise((resolve) => {
            const handler = (message) => {
                if (message.type === 'service_list') {
                    this.removeListener('service_list', handler);
                    const services = message.services.map(s => new ServiceSummary(s));
                    resolve(services);
                }
            };
            this.on('service_list', handler);
        });
    }
    
    /**
     * Get current subscriptions
     * 
     * @returns {string[]}
     */
    getSubscriptions() {
        return [...this.subscriptions];
    }
}

// ============================================================================
// Example Usage
// ============================================================================

/**
 * Basic usage example
 */
async function exampleBasic() {
    console.log('=== Basic WebSocket Example ===\n');
    
    const client = new SongbirdWebSocketClient('ws://localhost:8080/api/ws/ws');
    
    try {
        // Connect
        await client.connect();
        console.log('✅ Connected to Songbird\n');
        
        // Query status
        const status = await client.queryStatus();
        console.log('📊 Federation Status:');
        console.log(`   Services: ${status.totalServices}`);
        console.log(`   Peers: ${status.totalPeers}`);
        console.log(`   Uptime: ${status.uptimeSeconds}s\n`);
        
        // Close
        await client.close();
        console.log('✅ Connection closed');
        
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

/**
 * Event subscription example
 */
async function exampleSubscriptions() {
    console.log('=== Event Subscription Example ===\n');
    
    const client = new SongbirdWebSocketClient('ws://localhost:8080/api/ws/ws');
    
    try {
        await client.connect();
        console.log('✅ Connected to Songbird\n');
        
        // Subscribe to events
        await client.subscribe(['service_update', 'health_update']);
        console.log('✅ Subscribed to events\n');
        
        // Listen for events
        console.log('🔊 Listening for events (Ctrl+C to stop)...\n');
        
        client.on('service_update', (event) => {
            console.log('📨 Service Update:');
            console.log(`   Service: ${event.service_name}`);
            console.log(`   Status: ${event.status}`);
            console.log(`   Address: ${event.address}\n`);
        });
        
        client.on('health_update', (event) => {
            console.log('📨 Health Update:');
            console.log(`   Service: ${event.service_name}`);
            console.log(`   Healthy: ${event.healthy}`);
            if (event.message) {
                console.log(`   Message: ${event.message}`);
            }
            console.log();
        });
        
        // Keep running for 30 seconds
        await new Promise(resolve => setTimeout(resolve, 30000));
        
        await client.close();
        
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

/**
 * Service discovery example
 */
async function exampleServiceDiscovery() {
    console.log('=== Service Discovery Example ===\n');
    
    const client = new SongbirdWebSocketClient('ws://localhost:8080/api/ws/ws');
    
    try {
        await client.connect();
        console.log('✅ Connected to Songbird\n');
        
        // Query services with ML capability
        console.log('🔍 Querying services with \'ml\' capability...');
        const services = await client.queryServices(['ml']);
        
        console.log(`📋 Found ${services.length} service(s):\n`);
        services.forEach(service => {
            console.log(`   • ${service.name}`);
            console.log(`     Address: ${service.address}:${service.port}`);
            console.log(`     Capabilities: ${service.capabilities.join(', ')}\n`);
        });
        
        await client.close();
        
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

/**
 * Run examples
 */
async function main() {
    console.log('\n' + '='.repeat(80));
    console.log('Songbird WebSocket Client - JavaScript/Node.js Examples');
    console.log('='.repeat(80) + '\n');
    
    // Run examples
    await exampleBasic();
    console.log('\n' + '-'.repeat(80) + '\n');
    
    await exampleServiceDiscovery();
    console.log('\n' + '-'.repeat(80) + '\n');
    
    // Uncomment to test event subscriptions (requires active events)
    // await exampleSubscriptions();
}

// Export for use as a module
module.exports = {
    SongbirdWebSocketClient,
    FederationStatus,
    ServiceSummary
};

// Run examples if executed directly
if (require.main === module) {
    main().catch(console.error);
}

File: examples/clients/python/websocket_client.py
#!/usr/bin/env python3
"""
Songbird WebSocket Client - Real-Time Communication

Official Python client for Songbird WebSocket API.
Provides real-time bidirectional communication with event subscriptions.

Version: 0.2.1
Last Updated: November 11, 2025 - Phase 4

Features:
- Real-time bidirectional communication
- Event subscription system
- Automatic reconnection
- Ping/pong keep-alive
- Query federation status
- Service discovery
- Type hints and full error handling

Requirements:
    pip install websockets asyncio

Usage:
    # Basic connection
    client = SongbirdWebSocketClient("ws://localhost:8080/api/ws/ws")
    await client.connect()
    
    # Subscribe to events
    await client.subscribe(["service_update", "health_update"])
    
    # Query status
    status = await client.query_status()
    
    # Query services
    services = await client.query_services(["ml", "training"])
    
    # Listen for events
    async for event in client.listen():
        print(f"Event: {event}")
"""

import asyncio
import json
import logging
from typing import Any, Dict, List, Optional, AsyncIterator
from dataclasses import dataclass
from datetime import datetime

try:
    import websockets
    from websockets.client import WebSocketClientProtocol
except ImportError:
    print("Error: websockets not installed. Run: pip install websockets")
    raise

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger('songbird.websocket')


@dataclass
class FederationStatus:
    """Federation status information"""
    total_services: int
    total_peers: int
    uptime_seconds: int
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'FederationStatus':
        return cls(
            total_services=data['total_services'],
            total_peers=data['total_peers'],
            uptime_seconds=data['uptime_seconds']
        )


@dataclass
class ServiceSummary:
    """Service summary information"""
    name: str
    address: str
    port: int
    capabilities: List[str]
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ServiceSummary':
        return cls(
            name=data['name'],
            address=data['address'],
            port=data['port'],
            capabilities=data['capabilities']
        )


class SongbirdWebSocketError(Exception):
    """Base exception for Songbird WebSocket errors"""
    pass


class ConnectionError(SongbirdWebSocketError):
    """Connection-related errors"""
    pass


class MessageError(SongbirdWebSocketError):
    """Message parsing/handling errors"""
    pass


class SongbirdWebSocketClient:
    """
    Songbird WebSocket Client
    
    Provides real-time bidirectional communication with Songbird.
    
    Example:
        client = SongbirdWebSocketClient("ws://localhost:8080/api/ws/ws")
        await client.connect()
        
        # Subscribe to events
        await client.subscribe(["service_update", "health_update"])
        
        # Query status
        status = await client.query_status()
        print(f"Services: {status.total_services}, Peers: {status.total_peers}")
        
        # Listen for events
        async for event in client.listen():
            if event['type'] == 'service_update':
                print(f"Service updated: {event['service_name']}")
        
        await client.close()
    """
    
    def __init__(
        self,
        url: str,
        ping_interval: float = 30.0,
        ping_timeout: float = 10.0,
        auto_reconnect: bool = True,
        max_reconnect_attempts: int = 5
    ):
        """
        Initialize WebSocket client
        
        Args:
            url: WebSocket URL (e.g., "ws://localhost:8080/api/ws/ws")
            ping_interval: Seconds between ping messages
            ping_timeout: Timeout for pong response
            auto_reconnect: Automatically reconnect on connection loss
            max_reconnect_attempts: Maximum reconnection attempts
        """
        self.url = url
        self.ping_interval = ping_interval
        self.ping_timeout = ping_timeout
        self.auto_reconnect = auto_reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        
        self._ws: Optional[WebSocketClientProtocol] = None
        self._connected = False
        self._ping_task: Optional[asyncio.Task] = None
        self._subscriptions: List[str] = []
        self._message_queue: asyncio.Queue = asyncio.Queue()
        self._reconnect_attempts = 0
        
    async def connect(self) -> None:
        """
        Connect to Songbird WebSocket server
        
        Raises:
            ConnectionError: If connection fails
        """
        try:
            logger.info(f"Connecting to {self.url}")
            self._ws = await websockets.connect(
                self.url,
                ping_interval=self.ping_interval,
                ping_timeout=self.ping_timeout
            )
            self._connected = True
            self._reconnect_attempts = 0
            logger.info("Connected successfully")
            
            # Start ping task
            self._ping_task = asyncio.create_task(self._ping_loop())
            
            # Start message receiver
            asyncio.create_task(self._receive_messages())
            
        except Exception as e:
            logger.error(f"Connection failed: {e}")
            raise ConnectionError(f"Failed to connect: {e}")
    
    async def close(self) -> None:
        """Close WebSocket connection"""
        if self._ping_task:
            self._ping_task.cancel()
        
        if self._ws:
            await self._ws.close()
        
        self._connected = False
        logger.info("Connection closed")
    
    async def _ping_loop(self) -> None:
        """Send periodic ping messages"""
        while self._connected:
            try:
                await asyncio.sleep(self.ping_interval)
                await self.ping()
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.warning(f"Ping failed: {e}")
    
    async def _receive_messages(self) -> None:
        """Receive and queue messages"""
        try:
            async for message in self._ws:
                try:
                    data = json.loads(message)
                    await self._message_queue.put(data)
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to parse message: {e}")
        except websockets.exceptions.ConnectionClosed:
            self._connected = False
            if self.auto_reconnect:
                await self._reconnect()
    
    async def _reconnect(self) -> None:
        """Attempt to reconnect"""
        if self._reconnect_attempts >= self.max_reconnect_attempts:
            logger.error("Max reconnection attempts reached")
            return
        
        self._reconnect_attempts += 1
        delay = min(2 ** self._reconnect_attempts, 30)  # Exponential backoff
        
        logger.info(f"Reconnecting in {delay}s (attempt {self._reconnect_attempts})")
        await asyncio.sleep(delay)
        
        try:
            await self.connect()
            # Re-subscribe to events
            if self._subscriptions:
                await self.subscribe(self._subscriptions)
        except Exception as e:
            logger.error(f"Reconnection failed: {e}")
            await self._reconnect()
    
    async def _send_message(self, message: Dict[str, Any]) -> None:
        """Send message to server"""
        if not self._connected or not self._ws:
            raise ConnectionError("Not connected")
        
        try:
            await self._ws.send(json.dumps(message))
        except Exception as e:
            logger.error(f"Failed to send message: {e}")
            raise MessageError(f"Failed to send message: {e}")
    
    async def subscribe(self, events: List[str]) -> None:
        """
        Subscribe to event types
        
        Args:
            events: List of event types to subscribe to
                   (e.g., ["service_update", "health_update"])
        """
        self._subscriptions.extend(events)
        await self._send_message({
            "type": "subscribe",
            "events": events
        })
        logger.info(f"Subscribed to: {events}")
    
    async def unsubscribe(self, events: List[str]) -> None:
        """
        Unsubscribe from event types
        
        Args:
            events: List of event types to unsubscribe from
        """
        for event in events:
            if event in self._subscriptions:
                self._subscriptions.remove(event)
        
        await self._send_message({
            "type": "unsubscribe",
            "events": events
        })
        logger.info(f"Unsubscribed from: {events}")
    
    async def ping(self, data: Optional[str] = None) -> None:
        """
        Send ping message
        
        Args:
            data: Optional data to include in ping
        """
        message = {"type": "ping"}
        if data:
            message["data"] = data
        await self._send_message(message)
    
    async def query_status(self) -> FederationStatus:
        """
        Query federation status
        
        Returns:
            FederationStatus object
        """
        await self._send_message({"type": "query_status"})
        
        # Wait for response
        while True:
            event = await self._message_queue.get()
            if event.get('type') == 'federation_status':
                return FederationStatus.from_dict(event)
    
    async def query_services(self, capabilities: List[str]) -> List[ServiceSummary]:
        """
        Query services by capability
        
        Args:
            capabilities: Required capabilities
        
        Returns:
            List of ServiceSummary objects
        """
        await self._send_message({
            "type": "query_services",
            "capabilities": capabilities
        })
        
        # Wait for response
        while True:
            event = await self._message_queue.get()
            if event.get('type') == 'service_list':
                return [
                    ServiceSummary.from_dict(s)
                    for s in event.get('services', [])
                ]
    
    async def listen(self) -> AsyncIterator[Dict[str, Any]]:
        """
        Listen for events (async iterator)
        
        Yields:
            Event dictionaries
        
        Example:
            async for event in client.listen():
                if event['type'] == 'service_update':
                    print(f"Service: {event['service_name']}")
        """
        while self._connected:
            try:
                event = await asyncio.wait_for(
                    self._message_queue.get(),
                    timeout=1.0
                )
                yield event
            except asyncio.TimeoutError:
                continue
            except Exception as e:
                logger.error(f"Error receiving event: {e}")
                break
    
    @property
    def connected(self) -> bool:
        """Check if connected"""
        return self._connected
    
    @property
    def subscriptions(self) -> List[str]:
        """Get current subscriptions"""
        return self._subscriptions.copy()


# ============================================================================
# Example Usage
# ============================================================================

async def example_basic():
    """Basic usage example"""
    print("=== Basic WebSocket Example ===\n")
    
    client = SongbirdWebSocketClient("ws://localhost:8080/api/ws/ws")
    
    try:
        # Connect
        await client.connect()
        print("✅ Connected to Songbird\n")
        
        # Query status
        status = await client.query_status()
        print(f"📊 Federation Status:")
        print(f"   Services: {status.total_services}")
        print(f"   Peers: {status.total_peers}")
        print(f"   Uptime: {status.uptime_seconds}s\n")
        
        # Close
        await client.close()
        print("✅ Connection closed")
        
    except Exception as e:
        print(f"❌ Error: {e}")


async def example_subscriptions():
    """Event subscription example"""
    print("=== Event Subscription Example ===\n")
    
    client = SongbirdWebSocketClient("ws://localhost:8080/api/ws/ws")
    
    try:
        await client.connect()
        print("✅ Connected to Songbird\n")
        
        # Subscribe to events
        await client.subscribe(["service_update", "health_update"])
        print("✅ Subscribed to events\n")
        
        # Listen for 5 events
        print("🔊 Listening for events (Ctrl+C to stop)...\n")
        count = 0
        async for event in client.listen():
            print(f"📨 Event: {event.get('type')}")
            print(f"   Data: {event}")
            count += 1
            if count >= 5:
                break
        
        await client.close()
        
    except KeyboardInterrupt:
        print("\n⏹️  Stopped by user")
        await client.close()
    except Exception as e:
        print(f"❌ Error: {e}")


async def example_service_discovery():
    """Service discovery example"""
    print("=== Service Discovery Example ===\n")
    
    client = SongbirdWebSocketClient("ws://localhost:8080/api/ws/ws")
    
    try:
        await client.connect()
        print("✅ Connected to Songbird\n")
        
        # Query services with ML capability
        print("🔍 Querying services with 'ml' capability...")
        services = await client.query_services(["ml"])
        
        print(f"📋 Found {len(services)} service(s):\n")
        for service in services:
            print(f"   • {service.name}")
            print(f"     Address: {service.address}:{service.port}")
            print(f"     Capabilities: {', '.join(service.capabilities)}\n")
        
        await client.close()
        
    except Exception as e:
        print(f"❌ Error: {e}")


async def main():
    """Run examples"""
    print("\n" + "="*80)
    print("Songbird WebSocket Client - Python Examples")
    print("="*80 + "\n")
    
    # Run examples
    await example_basic()
    print("\n" + "-"*80 + "\n")
    
    await example_service_discovery()
    print("\n" + "-"*80 + "\n")
    
    # Uncomment to test event subscriptions (requires active events)
    # await example_subscriptions()


if __name__ == "__main__":
    asyncio.run(main())

